spark/src/main/scala/geotrellis/spark/io/tms/TMSReader.scala

```diff
@@ -0,0 +1,84 @@
+package geotrellis.spark.io.tms
+
+import geotrellis.raster._
+import geotrellis.raster.render._
+import geotrellis.spark.SpatialKey
+import geotrellis.vector.{Point, Extent}
+import geotrellis.spark.tiling.ZoomedLayoutScheme
+
+import spire.syntax.cfor._
+import scalaj.http._
+
+import scala.collection.mutable._
+import java.net.{URI, URL}
+import javax.imageio._
+
+
+class TMSReader[T](
+  uriTemplate: String,
+  f: URI => T = TMSReader.decodeTile _
+) extends Serializable {
+  def uri(z: Int, x: Int, y: Int): URI = new URI(
+    uriTemplate.replace("{z}", z.toString)
+      .replace("{x}", x.toString)
+      .replace("{y}", y.toString)
+  )
+
+  def read(zoom: Int, bounds: GridBounds): Iterator[(SpatialKey, T)] = {
+    // construct bounds from zoom + extent
+    for ((x, y) <- bounds.coordsIter)
+    yield SpatialKey(x, y) -> f(uri(zoom, x, y))
+  }
+
+  def read(zoom: Int, extent: Extent, tileSize: Int = ZoomedLayoutScheme.DEFAULT_TILE_SIZE): Iterator[(SpatialKey, T)] = {
+    val layoutDefinition = ZoomedLayoutScheme.layoutForZoom(zoom, extent, tileSize)
+    val bounds = layoutDefinition.mapTransform.extentToBounds(extent)
+    read(zoom, bounds)
+  }
+
+  def read(zoom: Int, x: Int, y: Int): T = f(uri(zoom, x, y))
+}
+
+object TMSReader {
+
+  def apply[T](uriTemplate: String, f: URI => T = TMSReader.decodeTile _): TMSReader[T] =
+    new TMSReader(uriTemplate, f)
+
+  // ImageIO use here to deal with different decoding behaviors (png vs jpg)
+  def decodeTile(uri: URI): Tile = {
+    val img = ImageIO.read(uri.toURL)
+    val height = img.getHeight()
+    val width = img.getWidth()
+    val rgbArr = img.getRGB(0, 0, width, height, null, 0, width)
+    IntArrayTile(rgbArr, width, height, None)
+  }
+
+  // ImageIO use here to deal with different decoding behaviors (png vs jpg)
+  def decodeMultibandTile(uri: URI): MultibandTile = {
+    val img = ImageIO.read(uri.toURL)
+    val height = img.getHeight()
+    val width = img.getWidth()
+    val length = width * height
+
+    val rgbArr = img.getRGB(0, 0, width, height, null, 0, width)
```

Comment:
It is two nested loops:

```java
for (int y = startY; y < startY+h; y++, yoff+=scansize) {
  off = yoff;
  for (int x = startX; x < startX+w; x++) {
    rgbArray[off++] = colorModel.getRGB(raster.getDataElements(x, y, data));
   }
}
```

Would not it be more efficient instead of doing it just to have two nested loops and to get pixels by col / row directly from memory without this extra copy step?

```scala
cfor(0)(_ < length, _ + 1) { c =>
  cfor(0)(_ < length, _ + 1) { r =>
  val c = img.getRGB(c, r)
  reds(cellIdx) = cell.red.toByte
  // ...
  }
}
```

Reply:
I vote for using `cfor` just for consistency with other loops in GT.